source/common/filesystem/posix/filesystem_impl.cc

```diff
@@ -126,16 +126,26 @@ bool InstanceImplPosix::illegalPath(const std::string& path) {
     return true;
   }
 
+  return illegalCanonicalPath(canonical_path.rc_);
+}
+
+bool InstanceImplPosix::illegalCanonicalPath(const absl::string_view& canonical) {
   // Platform specific path sanity; we provide a convenience to avoid Envoy
   // instances poking in bad places. We may have to consider conditioning on
   // platform in the future, growing these or relaxing some constraints (e.g.
   // there are valid reasons to go via /proc for file paths).
-  // TODO(htuch): Optimize this as a hash lookup if we grow any further.
-  if (absl::StartsWith(canonical_path.rc_, "/dev") ||
-      absl::StartsWith(canonical_path.rc_, "/sys") ||
-      absl::StartsWith(canonical_path.rc_, "/proc")) {
-    return true;
+  auto& paths = []() -> const std::vector<absl::string_view>& {
+    CONSTRUCT_ON_FIRST_USE(std::vector<absl::string_view>, {"/dev", "/sys", "/proc"});
+  }();
```

Comment:
Sorry for the back and forth on this, but this lambda doesn't seem to provide any real utility. Our normal pattern is a separate static method. If you really don't want to do that, my reading of the [style guide](https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables) says you can do this:

```
ABSL_CONST_INIT static constexpr std::array<absl::string_view, 3> paths = {{"/dev", "/sys", "/proc"}};
```

My reasoning is that std::array and absl::string_view are trivially destructible (and therefore destructor execution is not subject to ordering). Further, absl::string_view(const char*) and the std::array initializer are constant expressions. Thus, this static is allowed by both tests (decision on destruction and decision on initialization) given in the style guide.